### PR TITLE
Improve the `Interface` trait

### PR DESCRIPTION
**Pull request by rylev**
*Note*: since this involves changes to generated code, it's easiest to review this commit by commit. The codegen was run in a separate commit from the other changes.

This improves the safety auditability of the `Interface` trait:
* Adds safety docs to the `Interface` trait. Even though this trait is not meant for public consumption, the docs can really help those reviewing the safety of generated code.
* Adds an `as_raw` method to the trait which simply wraps `std::mem::transmute_copy`. `transmute_copy` is very powerful and easy to use wrong. Having a bespoke method for converting an interface to a raw pointer which is documented as such makes the code easier to read. We've also had several [users ask for this type of functionality before](https://github.com/microsoft/windows-rs/issues/1556). 
* Unmarks `Interface::vtable` as `unsafe` since this method is in fact safe. 

Finally, we use the `Interface::as_raw` method instead of `std::mem::transmute_copy` in the generated code.  

cc @eholk

### CODE REVIEW

**Comment by kennykerr** [comment]
> Adds safety docs to the Interface trait. Even though this trait is not meant for public consumption, the docs can really help those reviewing the safety of generated code.

Note that parts of this trait are meant for public consumption. That's why some members were hidden. Notably, `IID` and `cast` are meant to be used by developers. The rest should remain hidden.

**Comment by rylev** [comment]
@kennykerr that sounds fine. I would argue for the newly added `as_raw` to also be publicly accessible as I think it's a better API (with many parallels in the standard library) than having to use `std::mem::transmute_copy`.

File: crates/libs/windows/src/core/interface.rs [hunk 1]
```diff
@@ -7,42 +7,67 @@ use bindings::*;
...
+///
+/// It is only safe to implement this trait if the implementing type is a valid COM interface pointer meaning
+/// the following criteria are met:
+/// * its in-memory representation is equal to `NonNull<NonNull<Self::VTable>>`
+/// * the vtable is correctly structured beginning with the `IUnknown` function pointers
+/// * the vtable must be the correct vtable for the supplied IID
 pub unsafe trait Interface: Sized {
-    #[doc(hidden)]
+    /// The interface's vtable
     type Vtable;
```
Comment (kennykerr) [inline crates/libs/windows/src/core/interface.rs:18]: Let's keep this hidden.

**Review by kennykerr — DISMISSED** [review]

**Review by eholk — DISMISSED** [review]

**Comment by rylev** [comment]
@kennykerr the latest nightly of clippy brought some new things to fix which I did, but it made the reviews stale. Can you re-review (only the last commit is new), and merge?

**Review by kennykerr — APPROVED** [review]

**Pull request merged**